pkg/virt-handler/node-labeller/node_labeller.go

```diff
@@ -0,0 +1,283 @@
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package nodelabeller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+	virtconfig "kubevirt.io/kubevirt/pkg/virt-config"
+	device_manager "kubevirt.io/kubevirt/pkg/virt-handler/device-manager"
+)
+
+const (
+	DeprecatedLabelNamespace              = "feature.node.kubernetes.io"
+	DeprecatedLabellerNamespaceAnnotation = "node-labeller-feature.node.kubernetes.io"
+	kvmPath                               = "/dev/kvm"
+	DeprecatedcpuFeaturePrefix            = "/cpu-feature-"
+	DeprecatedcpuModelPrefix              = "/cpu-model-"
+	DeprecatedHyperPrefix                 = "/kvm-info-cap-hyperv-"
+	CPUFeaturePrefix                      = "cpu-feature.node.kubevirt.io/"
+	cpuModelPrefix                        = "cpu-model.node.kubevirt.io/"
+	hypervPrefix                          = "hyperv.node.kubevirt.io/"
+)
+
+//NodeLabeller struct holds informations needed to run node-labeller
+type NodeLabeller struct {
+	kvmController    *device_manager.DeviceController
+	nodeInformer     cache.SharedIndexInformer
+	clientset        kubecli.KubevirtClient
+	queue            workqueue.RateLimitingInterface
+	host             string
+	namespace        string
+	logger           *log.FilteredLogger
+	clusterConfig    *virtconfig.ClusterConfig
+	hypervFeatures   supportedFeatures
+	hostCapabilities supportedFeatures
+	cpuInfo          cpuInfo
+}
+
+func NewNodeLabeller(kvmController *device_manager.DeviceController, clusterConfig *virtconfig.ClusterConfig, nodeInformer cache.SharedIndexInformer, clientset kubecli.KubevirtClient, host, namespace string) *NodeLabeller {
+	return &NodeLabeller{
+		kvmController: kvmController,
+		nodeInformer:  nodeInformer,
+		clientset:     clientset,
+		queue:         workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		host:          host,
+		namespace:     namespace,
+		logger:        log.DefaultLogger(),
+		clusterConfig: clusterConfig,
+	}
+}
+
+//Run runs node-labeller
+func (n *NodeLabeller) Run(threadiness int, stop chan struct{}) {
+	defer n.queue.ShutDown()
+	n.logger.Infof("node-labeller is running")
+
+	err := n.loadCPUInfo()
+	if err != nil {
+		n.logger.Errorf("node-labeller could not load cpu info: " + err.Error())
+	}
+
+	err = n.loadHostCapabilities()
+	if err != nil {
+		n.logger.Errorf("node-labeller could not load host capabilities: " + err.Error())
+	}
+
+	n.loadHypervFeatures()
+
+	n.clusterConfig.SetConfigModifiedCallback(func() {
+		err := n.run()
+		if err != nil {
+			n.logger.Errorf("node-labeller cannot update labels %s", err.Error())
+		}
```

Comment:
we don't want to perform work in a callback like this. instead, replace this call to `n.run()` with  `n.queue.Add(n.host)`. by queueing the key we are ensuring that the key will be worked on by the reconcile loop.